### PR TITLE
feat: the law of leaky abstractions

### PR DESCRIPTION
**Pull request by dwmkerr**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -180,6 +181,26 @@ Some complexity in a system is 'inadvertent'. It is a consequence of poor struct
...
 
+### The Law of Leaky Abstractions
+
+[The Law of Leaky Abstractions on Joel on Software](https://www.joelonsoftware.com/2002/11/11/the-law-of-leaky-abstractions/)
+
+> All non-trivial abstractions, to some degree, are leaky.
+>
+> (Joel Spolsky)
+
+This law states that abstractions, which are generally used in computing to simplify working with complicated systems, will in certain situations 'leak' elements of the underlying system, this making the abstraction behave in an unexpected way.
```
Comment (prabugp-alt) [inline README.md]: Did you mean to say 'thus making the abstraction behave...'?

File: README.md [hunk 2]
```diff
@@ -180,6 +181,26 @@ Some complexity in a system is 'inadvertent'. It is a consequence of poor struct
...
+
+An example might be loading a file and reading its contents. The file system APIs are an _abstraction_ of the lower level kernel systems, which are themselves an abstraction over the physical processes relating to changing data on a magnetic platter or flash memory for an SSD. In most cases, the abstraction of treating a file like a stream of binary data will work. But if the physical drive is damaged, the results of reading the data may be undefined.
```
Comment (prabugp-alt) [inline README.md]: Not completely clear on how an undefined behaviour is the same as leaky abstraction. Could it be construed as bad abstraction without specifying error cases? How do we differentiate between the two? 

The way one could see this example is - to handle the error case, one has to understand the underlying implementation details which is not explicit while using the abstraction itself. Hence the leak is non-obvious and probably explain why it is hard to provide an abstraction for this scenario without leaking anything.

File: README.md (hunk 1 above)
Comment (dwmkerr) [reply README.md]: Well spotted, fixed, thank-you

File: README.md (hunk 2 above)
Comment (dwmkerr) [reply README.md]: I've tried to replace this with a better example, WDYT?

**Comment by prabugp-alt** [comment]
LGTM. 👍

**Pull request merged**